elastalert/kibana_discover.py

```diff
@@ -0,0 +1,192 @@
+    from_time = ts_add(timestamp, -from_timedelta)
+    to_timedelta = rule.get('kibana_discover_to_timedelta', timeframe)
+    to_time = ts_add(timestamp, to_timedelta)
+
+    if kibana_version in kibana5_kibana6_versions:
+        globalState = kibana6_disover_global_state(from_time, to_time)
+        appState = kibana_discover_app_state(index, columns, filters, query_keys, match)
+
+    elif kibana_version in kibana7_versions:
+        globalState = kibana7_disover_global_state(from_time, to_time)
+        appState = kibana_discover_app_state(index, columns, filters, query_keys, match)
+
+    else:
+        logging.warning(
+            'Unknown kibana discover application version %s for rule %s' % (
+                kibana_version,
+                rule.get('name', '<MISSING NAME>')
+            )
+        )
+        return None
+
+    return "%s?_g=%s&_a=%s" % (
+        os.path.expandvars(discover_url),
+        urllib.parse.quote(globalState),
+        urllib.parse.quote(appState)
+    )
+
+
+def kibana6_disover_global_state(from_time, to_time):
+    return prison.dumps( {
+        'refreshInterval': {
+            'pause': True,
+            'value': 0
+        },
+        'time': {
+            'from': from_time,
+            'mode': 'absolute',
+            'to': to_time
+        }
+    } )
+
+
+def kibana7_disover_global_state(from_time, to_time):
+    return prison.dumps( {
+        'filters': [],
+        'refreshInterval': {
+            'pause': True,
+            'value': 0
+        },
+        'time': {
+            'from': from_time,
+            'to': to_time
+        }
+    } )
+
+
+def kibana_discover_app_state(index, columns, filters, query_keys, match):
+    app_filters = []
+
+    if filters:
+        bool_filter = { 'must': filters }
+        app_filters.append( {
+            '$state': {
+                'store': 'appState'
+            },
+            'bool': bool_filter,
+            'meta': {
+                'alias': 'filter',
+                'disabled': False,
+                'index': index,
+                'key': 'bool',
+                'negate': False,
+                'type': 'custom',
+                'value': json.dumps(bool_filter, separators=(',', ':'))
+            },
+        } )
+
+    for query_key in query_keys:
+        query_value = lookup_es_key(match, query_key)
+
+        if query_value is None:
+            app_filters.append( {
+                '$state': {
+                    'store': 'appState'
+                },
+                'exists': {
+                    'field': query_key
+                },
+                'meta': {
+                    'alias': None,
+                    'disabled': False,
+                    'index': index,
+                    'key': query_key,
+                    'negate': True,
+                    'type': 'exists',
+                    'value': 'exists'
+                }
+            } )
+
+        else:
```

Comment:
Filters for query keys are based on the same behaviour as the add filter button.

![image](https://user-images.githubusercontent.com/12101120/65259153-35ac9180-dad2-11e9-8d41-8675e67ff7bd.png)

![image](https://user-images.githubusercontent.com/12101120/65259775-3d206a80-dad3-11e9-915a-8236f5d769ad.png)

Using Kibana's Sample web logs:

```
name: Frequent Responses
index: kibana_sample_data_logs

timestamp_field: "timestamp"
timestamp_type: "custom"
timestamp_format: "%Y-%m-%dT%H:%M:%S.%fZ"

type: frequency
num_events: 3

timeframe:
  minutes: 10

query_key: response

use_kibana_discover: true
kibana_discover_url: http://localhost:5601/#/discover
kibana_discover_version: '6.8'
kibana_discover_index_pattern_id: 90943e30-9a47-11e8-b64d-95841ca0b247 # specific to my instance
kibana_discover_columns: [ 'message', 'request', 'response' ]

alert: [ debug ]
alert_text: '{0}'
alert_text_args: [ kibana_link ]
alert_text_type: alert_text_only
```